_posts/2020-01-14-role-assumption-tricks.md

```diff
@@ -0,0 +1,131 @@
+        }
+      ]
+    }
+   ```
+
+Imagine we were to assume a role in another account to access S3 from within that context. The following code will assume role using the javascript SDK for this scenario, and provide those credentials to the S3 account. Using plain STS client calls, it looks like the following;
+
+```javascript
+import { S3, STS, Credentials } from "aws-sdk";
+
+const { 
+  Credentials: { 
+    AccessKeyId: accessKeyId, 
+    SecretAccessKey: secretAccessKey,
+    SessionToken: sessionToken 
+  } 
+} = await new STS().assumeRole({
+  RoleArn: "arn:aws:iam::0987654321:role/role-to-assume"
+}).promise();
+
+const client = new S3({
+  credentials: new Credentials({ accessKeyId, secretAccessKey, SessionToken })
+});
+```
+
+There's obviously a lot of boilerplate here, primarily because of the changing case of the input and output parameters between the response of the STS call and the credentials object. Removing this boilerplate was my reason for writing my own helper library in the first place. Now that ChainableTemporaryCredentials is around, we get rid of some the ceremony. Check this out;
+
+```typescript
+import { S3, ChainableTemporaryCredentials } from "aws-sdk";
+
+const credentials = new ChainableTemporaryCredentials({
+  params: {
+    // Any parameters used by STS AssumeRole can be used here eg; RoleSessionName etc
+    RoleArn: "arn:aws:iam::0987654321:role/role-to-assume"
+  }
+});
+
+const client = new S3({ credentials });
+```
+
+# Role Chaining
+
+Extending this to a third role that is assumable from a 'middle' role isn't an awful lot different from the example with two roles. We simply add another role, and place a trust on the role in the middle.
+
+<image showing the trusts>
+
+Using ChainableTemporaryCredentials we can perform the double-assumption by adding an additional parameter. 'masterCredentials' can be used to specify how the credentials to the top level call should be acquired.
+
+```typescript
+import { S3, ChainableTemporaryCredentials } from "aws-sdk";
+
+const credentials = new ChainableTemporaryCredentials({
+  params: {
+    RoleArn: "arn:aws:iam::0101010101:role/next-role-to-assume"
+  },
+  masterCredentials: new AWS.ChainableTemporaryCredentials({
+    params: { 
+      RoleArn: "arn:aws:iam::0987654321:role/role-to-assume"
+    }
+  })
+});
+
+const client = new S3({ credentials });
+```
+
+Simples! You can probably imagine how ugly it gets when directly using STS calls, hence why I wrote my own library to handle it!
+
+# Conclusion
+
+We had a brief look at how cross-account role assumption works, and how to set it up in the simple two account case. We showed how to do this STS calls, and how the ChainableTemporaryCredentials provider in Javascript SDK makes this easier. Then we added a third role, and showed how to perform role chaining via the credential provider. We gained an appreciation for how this makes the entire process simpler!
```

Comment:
this STS calls, and how the ChainableTemporaryCredentials

no need for comma after calls.